<a href="https://colab.research.google.com/github/alefs9/EAS_Aire_Limpio/blob/EDA/Exploraci%C3%B3n_del_dataset_EAS_Aire.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Configuración del Entorno

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
CARPETA = "/content/drive/MyDrive/Concurrente"

In [ ]:
!pip install duckdb -q

In [ ]:
import os
import duckdb
import pandas as pd

archivo = "/content/drive/MyDrive/Concurrente/EAS_Aire_0.csv"

con = duckdb.connect()

## Creación de una vista del dataset

In [ ]:
con.execute(f"""
CREATE OR REPLACE VIEW eas_aire AS

SELECT *
FROM read_csv_auto(
    '{archivo}',
    sample_size=100000
);
""")

print("Vista 'eas_aire' creada correctamente.")

Vista 'eas_aire' creada correctamente.


In [ ]:
muestra = con.execute("""
SELECT *
FROM eas_aire
LIMIT 5
""").df()

display(muestra)

,Número.de.informe,Nombre.de.la.Evaluación,Etapa,Componente.ambiental,Procedencia.de.la.Muestra,Procedencia.Especifica.de.la.Muestra,Nombre.del.punto,Este,Norte,Altitud,...,Tipo.de.muestra,Tipo.de.análisis,Periodo,Fecha.inicio,Hora.inicio,Fecha.fin,Hora.fin,Valor,Parámetro,Unidad.de.medida
0,REAS-006-2024-STEC,Calidad del aire en el ámbito de la zona indus...,No aplica,Aire,Aire ambiental,No aplica,CA-CH-03,767988,8992825,15,...,Simple,Concentración de gases,1 hora,44927,00:00:00,-,-,53.71,Dióxido de Azufre (SO2),µg/m³
1,REAS-006-2024-STEC,Calidad del aire en el ámbito de la zona indus...,No aplica,Aire,Aire ambiental,No aplica,CA-CH-03,767988,8992825,15,...,Simple,Concentración de gases,1 hora,44927,01:00:00,-,-,25.28,Dióxido de Azufre (SO2),µg/m³
2,REAS-006-2024-STEC,Calidad del aire en el ámbito de la zona indus...,No aplica,Aire,Aire ambiental,No aplica,CA-CH-03,767988,8992825,15,...,Simple,Concentración de gases,1 hora,44927,02:00:00,-,-,40.64,Dióxido de Azufre (SO2),µg/m³
3,REAS-006-2024-STEC,Calidad del aire en el ámbito de la zona indus...,No aplica,Aire,Aire ambiental,No aplica,CA-CH-03,767988,8992825,15,...,Simple,Concentración de gases,1 hora,44927,03:00:00,-,-,36.76,Dióxido de Azufre (SO2),µg/m³
4,REAS-006-2024-STEC,Calidad del aire en el ámbito de la zona indus...,No aplica,Aire,Aire ambiental,No aplica,CA-CH-03,767988,8992825,15,...,Simple,Concentración de gases,1 hora,44927,04:00:00,-,-,18.55,Dióxido de Azufre (SO2),µg/m³


In [ ]:
total_registros = con.execute("""
SELECT COUNT(*)
FROM eas_aire
""").fetchone()[0]

print(f"Registros totales: {total_registros:,}")

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

Registros totales: 11,407,394


## Dimensiones generales

In [ ]:
total_registros = con.execute("""
    SELECT COUNT(*)
    FROM eas_aire
""").fetchone()[0]

esquema = con.execute("""
    DESCRIBE eas_aire
""").df()

print(f"Registros totales: {total_registros:,}")
print(f"Cantidad de columnas: {len(esquema)}")

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

Registros totales: 11,407,394
Cantidad de columnas: 23


## Estructura de las variables

In [ ]:
display(esquema)

,column_name,column_type,null,key,default,extra
0,Número.de.informe,VARCHAR,YES,None,None,None
1,Nombre.de.la.Evaluación,VARCHAR,YES,None,None,None
2,Etapa,VARCHAR,YES,None,None,None
3,Componente.ambiental,VARCHAR,YES,None,None,None
4,Procedencia.de.la.Muestra,VARCHAR,YES,None,None,None
5,Procedencia.Especifica.de.la.Muestra,VARCHAR,YES,None,None,None
6,Nombre.del.punto,VARCHAR,YES,None,None,None
7,Este,BIGINT,YES,None,None,None
8,Norte,BIGINT,YES,None,None,None
9,Altitud,BIGINT,YES,None,None,None


## Muestra inicial de registros

In [ ]:
muestra = con.execute("""
    SELECT *
    FROM eas_aire
    LIMIT 10
""").df()

display(muestra)

,Número.de.informe,Nombre.de.la.Evaluación,Etapa,Componente.ambiental,Procedencia.de.la.Muestra,Procedencia.Especifica.de.la.Muestra,Nombre.del.punto,Este,Norte,Altitud,...,Tipo.de.muestra,Tipo.de.análisis,Periodo,Fecha.inicio,Hora.inicio,Fecha.fin,Hora.fin,Valor,Parámetro,Unidad.de.medida
0,REAS-006-2024-STEC,Calidad del aire en el ámbito de la zona indus...,No aplica,Aire,Aire ambiental,No aplica,CA-CH-03,767988,8992825,15,...,Simple,Concentración de gases,1 hora,44927,00:00:00,-,-,53.71,Dióxido de Azufre (SO2),µg/m³
1,REAS-006-2024-STEC,Calidad del aire en el ámbito de la zona indus...,No aplica,Aire,Aire ambiental,No aplica,CA-CH-03,767988,8992825,15,...,Simple,Concentración de gases,1 hora,44927,01:00:00,-,-,25.28,Dióxido de Azufre (SO2),µg/m³
2,REAS-006-2024-STEC,Calidad del aire en el ámbito de la zona indus...,No aplica,Aire,Aire ambiental,No aplica,CA-CH-03,767988,8992825,15,...,Simple,Concentración de gases,1 hora,44927,02:00:00,-,-,40.64,Dióxido de Azufre (SO2),µg/m³
3,REAS-006-2024-STEC,Calidad del aire en el ámbito de la zona indus...,No aplica,Aire,Aire ambiental,No aplica,CA-CH-03,767988,8992825,15,...,Simple,Concentración de gases,1 hora,44927,03:00:00,-,-,36.76,Dióxido de Azufre (SO2),µg/m³
4,REAS-006-2024-STEC,Calidad del aire en el ámbito de la zona indus...,No aplica,Aire,Aire ambiental,No aplica,CA-CH-03,767988,8992825,15,...,Simple,Concentración de gases,1 hora,44927,04:00:00,-,-,18.55,Dióxido de Azufre (SO2),µg/m³
5,REAS-006-2024-STEC,Calidad del aire en el ámbito de la zona indus...,No aplica,Aire,Aire ambiental,No aplica,CA-CH-03,767988,8992825,15,...,Simple,Concentración de gases,1 hora,44927,05:00:00,-,-,20.83,Dióxido de Azufre (SO2),µg/m³
6,REAS-006-2024-STEC,Calidad del aire en el ámbito de la zona indus...,No aplica,Aire,Aire ambiental,No aplica,CA-CH-03,767988,8992825,15,...,Simple,Concentración de gases,1 hora,44927,06:00:00,-,-,33.409999999999997,Dióxido de Azufre (SO2),µg/m³
7,REAS-006-2024-STEC,Calidad del aire en el ámbito de la zona indus...,No aplica,Aire,Aire ambiental,No aplica,CA-CH-03,767988,8992825,15,...,Simple,Concentración de gases,1 hora,44927,07:00:00,-,-,22.19,Dióxido de Azufre (SO2),µg/m³
8,REAS-006-2024-STEC,Calidad del aire en el ámbito de la zona indus...,No aplica,Aire,Aire ambiental,No aplica,CA-CH-03,767988,8992825,15,...,Simple,Concentración de gases,1 hora,44927,08:00:00,-,-,29.61,Dióxido de Azufre (SO2),µg/m³
9,REAS-006-2024-STEC,Calidad del aire en el ámbito de la zona indus...,No aplica,Aire,Aire ambiental,No aplica,CA-CH-03,767988,8992825,15,...,Simple,Concentración de gases,1 hora,44927,09:00:00,-,-,28.19,Dióxido de Azufre (SO2),µg/m³


## Cobertura temporal
La variable `Fecha.inicio` se encuentra almacenada como fecha serial de Excel.

Para interpretarla se utiliza como fecha base el 30 de diciembre de 1899. Luego se analiza
la cobertura temporal y la cantidad de registros disponibles para cada año.

In [ ]:
registros_por_anio = con.execute("""
WITH fechas AS (
    SELECT
        DATE '1899-12-30'
        + CAST(
            TRY_CAST("Fecha.inicio" AS DOUBLE)
            AS INTEGER
        ) AS fecha
    FROM eas_aire
)

SELECT
    YEAR(fecha) AS anio,
    COUNT(*) AS registros,
    MIN(fecha) AS fecha_minima,
    MAX(fecha) AS fecha_maxima
FROM fechas

GROUP BY YEAR(fecha)
ORDER BY anio
""").df()

display(registros_por_anio)

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,anio,registros,fecha_minima,fecha_maxima
0,2018,401514,2018-01-01,2018-12-31
1,2019,562484,2019-01-01,2019-12-31
2,2020,691601,2020-01-01,2020-12-31
3,2021,613407,2021-01-01,2021-12-31
4,2022,2784736,2022-01-01,2022-12-31
5,2023,3476018,2023-01-01,2023-12-31
6,2024,2877634,2024-01-01,2024-12-31


In [ ]:
rango_temporal = con.execute("""
WITH fechas AS (
    SELECT
        DATE '1899-12-30'
        + CAST(
            TRY_CAST("Fecha.inicio" AS DOUBLE)
            AS INTEGER
        ) AS fecha
    FROM eas_aire
)

SELECT
    MIN(fecha) AS fecha_minima,
    MAX(fecha) AS fecha_maxima
FROM fechas
""").df()

display(rango_temporal)

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,fecha_minima,fecha_maxima
0,2018-01-01,2024-12-31


## Registros pertenecientes al periodo reciente
Debido al criterio de actualidad establecido para el proyecto, se analiza cuántas observaciones
pertenecen al periodo 2022-2024.

En esta etapa únicamente se contabilizan; todavía no se eliminan registros anteriores.

In [ ]:
registros_recientes = con.execute("""
WITH fechas AS (
    SELECT
        DATE '1899-12-30'
        + CAST(
            TRY_CAST("Fecha.inicio" AS DOUBLE)
            AS INTEGER
        ) AS fecha
    FROM eas_aire
)

SELECT
    COUNT(*) AS registros_2022_2024
FROM fechas

WHERE fecha >= DATE '2022-01-01'
  AND fecha < DATE '2025-01-01'
""").df()

display(registros_recientes)

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,registros_2022_2024
0,9138388


## 9. Puntos de monitoreo

Se analiza la cantidad de puntos de monitoreo diferentes y la distribución de registros entre
ellos para identificar posibles diferencias de cobertura.

In [ ]:
cantidad_puntos = con.execute("""
SELECT
    COUNT(DISTINCT "Nombre.del.punto") AS puntos_unicos
FROM eas_aire
""").df()

display(cantidad_puntos)

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,puntos_unicos
0,151


In [ ]:
puntos = con.execute("""
SELECT
    "Nombre.del.punto" AS punto,
    COUNT(*) AS registros
FROM eas_aire

GROUP BY "Nombre.del.punto"
ORDER BY registros DESC
""").df()

display(puntos.head(30))

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,punto,registros
0,CA-TA-01,946479
1,CA-SB-01,642392
2,CA-CC-01,583730
3,CA-ILO-01,568384
4,CA-ILO-03,568384
5,CA-ILO-02,567904
6,CA-CH-03,401446
7,CA-CP-01,376412
8,CA-TO-01,334939
9,CA-RZC-01,315043


## 10. Tipos de análisis ambiental

Se identifica qué tipos de análisis componen el dataset y cuántas observaciones corresponden
a cada uno.

In [ ]:
tipos_analisis = con.execute("""
SELECT
    "Tipo.de.análisis" AS tipo_analisis,
    COUNT(*) AS registros

FROM eas_aire

GROUP BY "Tipo.de.análisis"
ORDER BY registros DESC
""").df()

display(tipos_analisis)

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,tipo_analisis,registros
0,Datos Meteorológicos,6812251
1,Concentración de gases,1574002
2,-,1330297
3,Concentración de partículas en PM10,857789
4,"Concentración de partículas en PM2,5",759264
5,Material particulado,36669
6,Concentración de metales en material particula...,26782
7,Metales en Filtro PM10,4091
8,Concentración de metales en PM10,3044
9,Metales Totales por ICP-OES,2848


## 11. Periodicidad de las mediciones

Las observaciones pueden corresponder a diferentes periodos de medición. Se analiza su
distribución para determinar si existen diferencias de granularidad temporal.

In [ ]:
periodos = con.execute("""
SELECT
    "Periodo" AS periodo,
    COUNT(*) AS registros

FROM eas_aire

GROUP BY "Periodo"
ORDER BY registros DESC
""").df()

display(periodos)

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,periodo,registros
0,1 hora,9446542
1,-,1134562
2,3 horas móviles,297011
3,8 horas móviles,201624
4,24 horas,128535
5,Móvil 8 horas,66048
6,3 horas móviles,59568
7,Móvil 3 horas,33024
8,Mensual,22689
9,3 horas moviles,17520


## 12. Parámetros ambientales disponibles

Se identifican los parámetros ambientales contenidos en el conjunto de datos, sus unidades
de medida y el número de registros asociados.

In [ ]:
parametros = con.execute("""
SELECT
    "Parámetro" AS parametro,
    "Unidad.de.medida" AS unidad,
    COUNT(*) AS registros

FROM eas_aire

GROUP BY
    "Parámetro",
    "Unidad.de.medida"

ORDER BY registros DESC
""").df()

display(parametros.head(100))

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,parametro,unidad,registros
0,Temperatura,°C,978995
1,Precipitación,mm,978995
2,Humedad Relativa,%,967451
3,Presión Barométrica,mmHg,731782
4,Dirección de Viento,°,712463
...,...,...,...
95,Silicio,µg/m3,753
96,Titanio,µg/m3,753
97,Vanadio,µg/m3,753
98,Zinc,µg/m3,753


## 13. Calidad de la variable de medición

La columna `Valor` fue detectada como texto debido a que no todas las observaciones contienen
un número.

Se determina cuántos valores pueden convertirse correctamente a valores numéricos y cuántos
requieren revisión.

In [ ]:
calidad_valor = con.execute("""
SELECT
    COUNT(*) AS total,

    COUNT(
        TRY_CAST(
            REPLACE(TRIM("Valor"), ',', '.')
            AS DOUBLE
        )
    ) AS numericos,

    COUNT(*) -
    COUNT(
        TRY_CAST(
            REPLACE(TRIM("Valor"), ',', '.')
            AS DOUBLE
        )
    ) AS no_numericos

FROM eas_aire
""").df()

display(calidad_valor)

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,total,numericos,no_numericos
0,11407394,8535307,2872087


## 14. Identificación de valores no numéricos

Se revisan las representaciones no numéricas presentes en la columna `Valor`.

El propósito es conocer su naturaleza antes de decidir cómo tratarlas durante la etapa
posterior de limpieza.

In [ ]:
valores_no_numericos = con.execute("""
SELECT
    TRIM("Valor") AS valor,
    COUNT(*) AS cantidad

FROM eas_aire

WHERE TRY_CAST(
    REPLACE(TRIM("Valor"), ',', '.')
    AS DOUBLE
) IS NULL

GROUP BY TRIM("Valor")
ORDER BY cantidad DESC
LIMIT 50
""").df()

display(valores_no_numericos)

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,valor,cantidad
0,-,1918259
1,NA,880299
2,CA,47628
3,MA,10650
4,N.D.,3798
5,<LC,3304
6,<L.C.,2605
7,FT,1268
8,< L.C.,1011
9,SE,705


## 15. Disponibilidad de valores numéricos por año

Se analiza la proporción de valores numéricos y no numéricos dentro del periodo reciente
2022-2024.

In [ ]:
calidad_anual = con.execute("""
WITH datos AS (
    SELECT
        DATE '1899-12-30'
        + CAST(
            TRY_CAST("Fecha.inicio" AS DOUBLE)
            AS INTEGER
        ) AS fecha,

        TRY_CAST(
            REPLACE(TRIM("Valor"), ',', '.')
            AS DOUBLE
        ) AS valor_numerico

    FROM eas_aire
)

SELECT
    YEAR(fecha) AS anio,
    COUNT(*) AS total,
    COUNT(valor_numerico) AS numericos,
    COUNT(*) - COUNT(valor_numerico) AS no_numericos,

    ROUND(
        100.0 * COUNT(valor_numerico) / COUNT(*),
        2
    ) AS porcentaje_numerico

FROM datos

WHERE fecha >= DATE '2022-01-01'
  AND fecha < DATE '2025-01-01'

GROUP BY YEAR(fecha)
ORDER BY anio
""").df()

display(calidad_anual)

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,anio,total,numericos,no_numericos,porcentaje_numerico
0,2022,2784736,2013381,771355,72.30
1,2023,3476018,2267167,1208851,65.22
2,2024,2877634,1994822,882812,69.32


## 16. Identificación de campos sin información útil

El dataset utiliza diferentes representaciones para indicar ausencia de información, como
guiones (`-`) o el texto `No aplica`.

Se revisan las principales variables para conocer la magnitud de estos casos.

In [ ]:
ausencias = con.execute("""
SELECT
    COUNT(*) AS total,

    COUNT(*) FILTER (
        WHERE TRIM(CAST("Etapa" AS VARCHAR)) = 'No aplica'
    ) AS etapa_no_aplica,

    COUNT(*) FILTER (
        WHERE TRIM(CAST("Procedencia.Especifica.de.la.Muestra" AS VARCHAR))
              = 'No aplica'
    ) AS procedencia_especifica_no_aplica,

    COUNT(*) FILTER (
        WHERE TRIM(CAST("Fecha.fin" AS VARCHAR)) = '-'
    ) AS fecha_fin_guion,

    COUNT(*) FILTER (
        WHERE TRIM(CAST("Hora.fin" AS VARCHAR)) = '-'
    ) AS hora_fin_guion,

    COUNT(*) FILTER (
        WHERE TRIM(CAST("Valor" AS VARCHAR)) = '-'
    ) AS valor_guion,

    COUNT(*) FILTER (
        WHERE TRIM(CAST("Tipo.de.análisis" AS VARCHAR)) = '-'
    ) AS tipo_analisis_guion,

    COUNT(*) FILTER (
        WHERE TRIM(CAST("Periodo" AS VARCHAR)) = '-'
    ) AS periodo_guion

FROM eas_aire
""").df()

display(ausencias)

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,total,etapa_no_aplica,procedencia_especifica_no_aplica,fecha_fin_guion,hora_fin_guion,valor_guion,tipo_analisis_guion,periodo_guion
0,11407394,11034974,11032061,10195881,10195881,1918259,1330297,1134562


## 17. Principales contaminantes atmosféricos

Se analiza la disponibilidad de mediciones de los principales contaminantes presentes en
el dataset durante el periodo 2022-2024.

Esta etapa tiene como objetivo evaluar su viabilidad para el posterior modelo de Machine
Learning, sin aplicar todavía filtros definitivos al archivo.

In [ ]:
contaminantes = con.execute("""
WITH base AS (
    SELECT
        TRIM("Parámetro") AS parametro,
        TRIM("Unidad.de.medida") AS unidad,
        TRIM("Periodo") AS periodo,

        DATE '1899-12-30'
        + CAST(
            TRY_CAST("Fecha.inicio" AS DOUBLE)
            AS INTEGER
        ) AS fecha,

        TRY_CAST(
            REPLACE(TRIM("Valor"), ',', '.')
            AS DOUBLE
        ) AS valor

    FROM eas_aire
),

clasificado AS (
    SELECT *,
        CASE

            WHEN (
                LOWER(parametro) LIKE '%particul%'
                OR LOWER(parametro) LIKE '%pm10%'
            )
            AND (
                LOWER(parametro) LIKE '%10%'
                OR LOWER(parametro) LIKE '%pm10%'
            )
            AND NOT (
                LOWER(parametro) LIKE '%2,5%'
                OR LOWER(parametro) LIKE '%2.5%'
                OR LOWER(parametro) LIKE '%pm2%'
            )
            THEN 'PM10'

            WHEN
                LOWER(parametro) LIKE '%2,5%'
                OR LOWER(parametro) LIKE '%2.5%'
                OR LOWER(parametro) LIKE '%pm2,5%'
                OR LOWER(parametro) LIKE '%pm2.5%'
            THEN 'PM2.5'

            WHEN LOWER(parametro) LIKE '%dióxido de azufre%'
            THEN 'SO2'

            WHEN LOWER(parametro) LIKE '%dióxido de nitrógeno%'
            THEN 'NO2'

            WHEN LOWER(parametro) LIKE '%monóxido de carbono%'
            THEN 'CO'

            WHEN LOWER(parametro) LIKE '%ozono%'
            THEN 'O3'

            ELSE NULL

        END AS contaminante

    FROM base
)

SELECT
    contaminante,
    unidad,
    periodo,
    COUNT(*) AS registros

FROM clasificado

WHERE fecha >= DATE '2022-01-01'
  AND fecha < DATE '2025-01-01'
  AND valor IS NOT NULL
  AND contaminante IS NOT NULL

GROUP BY
    contaminante,
    unidad,
    periodo

ORDER BY
    contaminante,
    registros DESC
""").df()

display(contaminantes)

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,contaminante,unidad,periodo,registros
0,CO,ppb,1 hora,77114
1,CO,µg/m³,1 hora,52569
2,CO,µg/m³,8 horas móviles,52536
3,CO,μg/m³,8 horas móviles,24572
4,CO,μg/m³,1 hora,24545
5,NO2,ppb,1 hora,85094
6,NO2,µg/m³,1 hora,58909
7,NO2,μg/m³,1 hora,26185
8,PM10,µg/m³,1 hora,326519
9,PM10,μg/m³,1 hora,140289


## 18. Mediciones horarias de los principales contaminantes

Se estudia específicamente la disponibilidad de mediciones con periodicidad de una hora,
ya que representan la mayor parte de las observaciones y proporcionan una granularidad
temporal homogénea.

In [ ]:
contaminantes_1h = contaminantes[
    contaminantes["periodo"]
        .str.strip()
        .str.lower()
        .eq("1 hora")
]

resumen_1h = (
    contaminantes_1h
    .groupby("contaminante", as_index=False)["registros"]
    .sum()
    .sort_values("registros", ascending=False)
)

display(resumen_1h)

print(
    "Total de mediciones horarias:",
    f"{resumen_1h['registros'].sum():,}"
)

## 19. Unidades utilizadas en los contaminantes

Se revisan las unidades de medida para identificar posibles diferencias de representación
o escalas incompatibles antes de diseñar el procedimiento de limpieza.

In [ ]:
unidades_contaminantes = (
    contaminantes_1h
    .groupby(
        ["contaminante", "unidad"],
        as_index=False
    )["registros"]
    .sum()
    .sort_values(
        ["contaminante", "registros"],
        ascending=[True, False]
    )
)

display(unidades_contaminantes)

# 20. Resumen de hallazgos

La exploración inicial permitió identificar las principales características del dataset:

- El conjunto contiene más de 11 millones de registros.
- La cobertura temporal comprende los años 2018 a 2024.
- El periodo 2022-2024 contiene más de 9 millones de registros.
- Existen 151 puntos de monitoreo.
- La mayoría de las mediciones posee periodicidad de una hora.
- La variable `Valor` contiene tanto valores numéricos como códigos y marcadores especiales.
- Se detectaron diferentes representaciones de ausencia de información.
- Existen diferencias de escritura en algunas categorías y unidades.
- Los principales contaminantes con suficiente información son PM10, PM2.5, SO2, NO2 y CO.
- Existen más de 1.6 millones de mediciones horarias numéricas de estos contaminantes durante 2022-2024.
- Los contaminantes están expresados en más de una unidad de medida, principalmente µg/m³ y ppb.

Estos resultados serán utilizados para definir y justificar las reglas de limpieza y
preprocesamiento en la siguiente etapa.